### PR TITLE
Geowave 615

### PR DESCRIPTION
**Pull request by dcy2003**
Fix for #615

SharedVisibilitySplittingIterator decomposes composite fieldIds on the tablet server, which replaces the old approach of doing this twice (once client side in AccumuloUtils and also in QueryFilterIterator)

FieldFilter filters only those fieldIds that were specified in QueryOptions

AttributesSubsetQueryIT tests both server side and client side filtering functionality

We also support subsetting through GeoTools Query via propertyNames.  GeoToolsAttributesSubsetTest verifies this functionality

### CODE REVIEW

File: core/store/src/main/java/mil/nga/giat/geowave/core/store/query/QueryOptions.java [hunk 1]
```diff
@@ -379,7 +407,27 @@ public static long getSerialversionuid() {
...
+					sBuf.append(comma);
+				}
+				sBuf.append(fieldId);
+			}
+			fieldIdsBytes = StringUtils.stringToBinary(sBuf.toString());
+			fieldIdsBytesNeeded = 4 + fieldIdsBytes.length;
+		}
+
+		final ByteBuffer buf = ByteBuffer.allocate(20 + authBytes.length + aSize + iSize + fieldIdsBytesNeeded);
+		buf.putInt((fieldIdsBytesNeeded > 0) ? 1 : 0);
```
Comment (rwgdrummer) [inline core/store/src/main/java/mil/nga/giat/geowave/core/store/query/QueryOptions.java]: Do we need the extra int, or can we just place a 0 if the field id list is emty?

File: extensions/adapters/vector/src/main/java/mil/nga/giat/geowave/adapter/vector/plugin/GeoWaveFeatureReader.java [hunk 2]
```diff
@@ -260,13 +263,17 @@ public BaseIssuer(
 		public CloseableIterator<SimpleFeature> query(
 				final PrimaryIndex index,
 				final mil.nga.giat.geowave.core.store.query.Query query ) {
+			final QueryOptions queryOptions = new QueryOptions(
```
Comment (rwgdrummer) [inline extensions/adapters/vector/src/main/java/mil/nga/giat/geowave/adapter/vector/plugin/GeoWaveFeatureReader.java]: Could simplifying this by having a new constructor that takes a List<String> for the subset attributes where getSubset() can return an empty list when subseting is not requested?

File: core/store/src/main/java/mil/nga/giat/geowave/core/store/query/QueryOptions.java [hunk 3]
```diff
@@ -379,7 +407,27 @@ public static long getSerialversionuid() {
 			}
 		}
 
-		final ByteBuffer buf = ByteBuffer.allocate(16 + authBytes.length + aSize + iSize);
+		int fieldIdsBytesNeeded = 0;
```
Comment (rwgdrummer) [inline core/store/src/main/java/mil/nga/giat/geowave/core/store/query/QueryOptions.java]: You used org.apache.commons.lang3.StringUtils.join else where.  Could we use that here?

File: extensions/datastores/accumulo/src/main/java/mil/nga/giat/geowave/datastore/accumulo/query/FieldFilter.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+			if (Arrays.equals(
+					k.getColumnQualifierData().getBackingArray(),
+					fieldId.getBytes())) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	public static IteratorSetting getIteratorSetting() {
```
Comment (rwgdrummer) [inline extensions/datastores/accumulo/src/main/java/mil/nga/giat/geowave/datastore/accumulo/query/FieldFilter.java]: Clean...we should use the approach else where.

File: extensions/datastores/accumulo/src/main/java/mil/nga/giat/geowave/datastore/accumulo/query/FieldFilter.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.accumulo.core.iterators.IteratorEnvironment;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+
+import com.google.common.base.Splitter;
+
+public class FieldFilter extends
+		Filter
+{
+	// must execute prior to WholeRowIterator/QueryFilterIterator but after
+	// SharedVisibilitySplittingIterator
```
Comment (rwgdrummer) [inline extensions/datastores/accumulo/src/main/java/mil/nga/giat/geowave/datastore/accumulo/query/FieldFilter.java]: Do we have an issue where subset fields may not include fields used in the filter?

File: extensions/datastores/accumulo/src/main/java/mil/nga/giat/geowave/datastore/accumulo/query/SharedVisibilitySplittingIterator.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+			throws IOException {
+
+		while (input.hasTop()) {
+
+			final Key currKey = input.getTopKey();
+			final Value currVal = input.getTopValue();
+
+			// check to see if the current value is a composite field containing
+			// multiple attributes with shared visibility
+			if (Arrays.equals(
```
Comment (rfecher) [inline extensions/datastores/accumulo/src/main/java/mil/nga/giat/geowave/datastore/accumulo/query/SharedVisibilitySplittingIterator.java]: we should go forward with this PR, but we will need to benchmark it at some point - particularly relating to the use of this transforming iterator (I'm not saying it'll turn out poor performance, just making sure we at some point quantify it).

**Comment by rwgdrummer** [comment]
Agreed.  Once we address these comments  We can move forward.

Steve and I were discussing potentially make a variant to the whole feature data adapter further reduce the volume from the server to the client.  

At some point, all of us did discuss a better method to compress the field/value bytes, using a bitmask--bits coincide with feature attribute names as they are in ordered within the simplefeaturetype). 

I think the next is benchmark this approach and then make the incremental changes to determine they value they provide.

File: core/store/src/main/java/mil/nga/giat/geowave/core/store/query/QueryOptions.java (hunk 1 above)
Comment (dcy2003) [inline core/store/src/main/java/mil/nga/giat/geowave/core/store/query/QueryOptions.java]: Don't need the extra int.  I will clean this up.

File: extensions/adapters/vector/src/main/java/mil/nga/giat/geowave/adapter/vector/plugin/GeoWaveFeatureReader.java (hunk 2 above)
Comment (dcy2003) [inline extensions/adapters/vector/src/main/java/mil/nga/giat/geowave/adapter/vector/plugin/GeoWaveFeatureReader.java]: Certainly could, though I'm personally of the opinion it's better to have constructors covering the required fields and setters for optional fields vs. a ton of permutations of constructors.

File: core/store/src/main/java/mil/nga/giat/geowave/core/store/query/QueryOptions.java (hunk 3 above)
Comment (dcy2003) [inline core/store/src/main/java/mil/nga/giat/geowave/core/store/query/QueryOptions.java]: Yep, will do.  Will be more concise.

File: extensions/datastores/accumulo/src/main/java/mil/nga/giat/geowave/datastore/accumulo/query/SharedVisibilitySplittingIterator.java (hunk 6 above)
Comment (dcy2003) [inline extensions/datastores/accumulo/src/main/java/mil/nga/giat/geowave/datastore/accumulo/query/SharedVisibilitySplittingIterator.java]: Agreed.  Said the exact same thing to Steve on Friday.  Will be curious to see what impact it has.

File: extensions/datastores/accumulo/src/main/java/mil/nga/giat/geowave/datastore/accumulo/query/FieldFilter.java (hunk 5 above)
Comment (dcy2003) [inline extensions/datastores/accumulo/src/main/java/mil/nga/giat/geowave/datastore/accumulo/query/FieldFilter.java]: I don't think we have an issue, but let's discuss to make sure I'm not missing something.

**Pull request merged**